### PR TITLE
entity discovery listeners [11093]

### PR DESCRIPTION
**Pull request by IkerLuengo**

### CODE REVIEW

**Review by EduPonz — CHANGES_REQUESTED** [review]
A rebase here would me much appreciated

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: src/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -29,6 +29,8 @@ set(${PROJECT_NAME}_SOURCES
     exception/Exception.cpp
     types/EntityId.cpp
     StatisticsBackend.cpp
+    subscriber/QosSerializer.cpp
+    subscriber/StatisticsParticipantListener.cpp
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/CMakeLists.txt:33]: Are they not?

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+
+            // Check whether the participant is already on the database
+            GUID_t participant_guid = info.info.m_guid;
+            auto participant_ids =
+                    database_->get_entities_by_guid(EntityKind::PARTICIPANT, to_string(participant_guid));
+            if (!participant_ids.empty())
+            {
+                throw Error("Participant discovered " + to_string(participant_guid)
+                              + " but there is already a Participant with the same GUID in the database");
+            }
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:221]: Do you agree removing then:
 - All checks for duplicate insertions
 - All checks for non-existent elements in the hierarchy (inserting a user on a host that is not in the database)
If `insert` is already checking this, I'm OK with it

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+            // Create the participant and push it to the queue
+            auto participant = std::make_shared<database::DomainParticipant>(
+                info.info.m_participantName.to_string(),
+                participant_info_to_backend_qos(info),
+                to_string(participant_guid),
+                std::shared_ptr<database::Process>(),
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:228]: Probably that works, as the nullptr will be casted to std::shared_ptr, but I believe this is formally more correct, given the constructor of the DomainParticipant

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        }
+        break;
+
+        case ParticipantDiscoveryInfo::CHANGED_QOS_PARTICIPANT:
+            // Update QoS on stored entity
+            break;
+
+        case ParticipantDiscoveryInfo::REMOVED_PARTICIPANT:
+        case ParticipantDiscoveryInfo::DROPPED_PARTICIPANT:
+            // Do nothing
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:241]: Moving this to a TODO

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+                to_string(participant_guid),
+                std::shared_ptr<database::Process>(),
+                domain);
+
+            entity_queue_->push(timestamp, participant);
+        }
+        break;
+
+        case ParticipantDiscoveryInfo::CHANGED_QOS_PARTICIPANT:
+            // Update QoS on stored entity
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:236]: Also moved to a TODO, waiting for the feature to be implemented

File: src/cpp/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/CMakeLists.txt:33]: The entire list is not (at least according to VSCode). I think it puts _ before . for instance.

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        DomainParticipant* participant,
+        ReaderDiscoveryInfo&& info)
+{
+    // First stop the data queue until the new entity is created
+    data_queue_->stop_consumer();
+
+    switch (info.status)
+    {
+        case ReaderDiscoveryInfo::DISCOVERED_READER:
+            process_endpoint_discovery(participant, info, EntityKind::DATAREADER, "DataReader");
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:261]: absolutely. That was a placeholder meant to be changed, and I forgot. I set the name as the GUID instead.

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+
+    // Check whether the locators are already in the database
+    for (auto dds_locator : info.info.remote_locators().unicast)
+    {
+        std::shared_ptr<database::Locator> locator;
+        auto locator_ids = database_->get_entities_by_name(EntityKind::LOCATOR, to_string(dds_locator));
+        if (locator_ids.empty())
+        {
+            // Create the Locator and push it to the queue
+            locator = std::make_shared<database::Locator>(to_string(dds_locator));
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp]: reworked the solution

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        std::static_pointer_cast<const database::Domain>(database_->get_entity(domain_ids.front().second)));
+
+    // Check whether the endpoint is already on the database
+    GUID_t endpoint_guid = info.info.guid();
+    auto endpoint_ids = database_->get_entities_by_guid(endpoint_kind, to_string(endpoint_guid));
+    if (!endpoint_ids.empty())
+    {
+        throw Error(endpoint_name + " discovered " + to_string(endpoint_guid)
+                      + " but there is already a " + endpoint_name + " with the same GUID in the database");
+    }
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:79]: removed

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+    // Get the domain from the database
+    auto domain_ids =
+            database_->get_entities_by_name(EntityKind::DOMAIN,
+                    std::to_string(statistics_participant->get_domain_id()));
+    if (domain_ids.empty())
+    {
+        throw Error(endpoint_name + " discovered on Domain " + std::to_string(statistics_participant->get_domain_id())
+                      + " but there is no such Domain in the database");
+    }
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:66]: We need the Domain in order to create the Topic. To get the existing Domain, the API makes us get the ID first and the Entity afterwards. In order to be safe, we need to check that the vector of IDs is not empty.

The same with the Participant and the Endpoint itself

File: src/cpp/subscriber/StatisticsParticipantListener.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp:247]: I found that we still need to get the Domain to create the Participant, and for that, we need to use `get_entities_by_name`, and to be sure, we have to check it does not return an empty vector.

**Review by EduPonz — CHANGES_REQUESTED** [review]
I'd remove the tests

statistics_participant_listener_tests.participant_qos_change_discovered
statistics_participant_listener_tests.reader_qos_change_discovered
statistics_participant_listener_tests.writer_qos_change_discovered

They fail because there is no implementation yet, but that' already on the list of pending tasks so we will not forget

File: src/cpp/subscriber/StatisticsParticipantListener.cpp [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        {
+            // Get the domain from the database
+            auto domain_ids =
+                    database_->get_entities_by_name(EntityKind::DOMAIN,
+                            std::to_string(participant->get_domain_id()));
+            if (domain_ids.empty())
+            {
+                throw Error("Participant discovered on Domain " + std::to_string(participant->get_domain_id())
+                              + " but there is no such Domain in the database");
+            }
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/subscriber/StatisticsParticipantListener.cpp]: Made a design decision: The statistics listeners for each monitor must be unique, and customized for that monitor in the sense that they receive the EntityId of the Domain for which they were created. As these listeners are instantiated by the backend itself, we can control this:
 - When the user creates a new monitor:
   - Create the Domain Entity first in the DB and get the given ID
   - Create the listener, with the given Domain ID
   - Create the statistics endpoints (participant, reader) with the listener

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: test/unittest/QosSerializer/QosSerializerTests.cpp [hunk 11]
```diff
@@ -0,0 +1,782 @@
...
+    qos.kind =  BY_SOURCE_TIMESTAMP_DESTINATIONORDER_QOS;
+    eprosima::statistics_backend::subscriber::serialize(qos, field, serialized);
+
+    expected[field.c_str()]["kind"] = "BY_SOURCE_TIMESTAMP_DESTINATIONORDER_QOS";
+    EXPECT_EQ(expected, serialized);
+}
+
+TEST(qos_serializer_tests, topic_data_qos_policy)
+{
+    TopicDataQosPolicy qos(PID_PAD);
```
Comment (JLBuenoLopez-eProsima) [inline test/unittest/QosSerializer/QosSerializerTests.cpp:214]: There is no need to pass `PID_PAD` as argument. It is enough with the default constructor. This also applies to the next two tests.

```
    TopicDataQosPolicy qos;
```

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
A couple of nitpicks.

However, I am concern about the tags being used both in the implementation and in the tests. Would it not be better to define those tags as constants in case that there is a later refactor in some of the QoS? I know it's a lot of work so maybe we can proceed as it is and improve later on.

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM with green CI (though the coverage work will probably complain about the updates not yet implemented).

**Pull request merged**